lib/neurow/jwt_auth_plug.ex

```diff
@@ -0,0 +1,136 @@
+
+                  {:error, code, message} ->
+                    conn |> forbidden(code, message)
+                end
+
+              {:error, code, message} ->
+                conn |> forbidden(code, message)
+            end
+
+          {:error, code, message} ->
+            conn |> forbidden(code, message)
+        end
+
+      {:error, code, message} ->
+        conn |> forbidden(code, message)
+    end
+  end
+
+  defp jwt_token_from_request(conn) do
+    case conn |> get_req_header("authorization") do
+      ["Bearer " <> jwt_token] ->
+        {:ok, jwt_token}
+
+      _ ->
+        {:error, :invalid_authorization_header, "Invalid authorization header"}
+    end
+  end
+
+  defp parse_jwt_token(jwt_token_str) do
+    try do
+      protected = JOSE.JWT.peek_protected(jwt_token_str)
+      payload = JOSE.JWT.peek_payload(jwt_token_str)
+      {:ok, protected, payload}
+    rescue
+      _ in Jason.DecodeError -> {:error, :invalid_jwt_token, "Invalid JWT token"}
+      _ in ArgumentError -> {:error, :invalid_jwt_token, "Invalid JWT token"}
+    end
+  end
+
+  defp fetch_jwk_from_issuer(payload, options) do
+    case payload do
+      %JOSE.JWT{fields: %{"iss" => issuer}} ->
+        jwk = options[:jwk_provider].(issuer)
+
+        if jwk != nil do
+          {:ok, jwk}
+        else
+          {:error, :unkown_issuer, "Unknown issuer"}
+        end
+
+      _ ->
+        {:error, :missing_iss_claim, "Mising iss claim"}
+    end
+  end
+
+  defp check_signature(jwt_token_str, jwk, options) do
+    case JOSE.JWT.verify_strict(jwk, [options[:allowed_algorithm]], jwt_token_str) do
+      {true, _jwt, _jws} -> {:ok}
+      {false, _jwt, _jws} -> {:error, :invalid_signature, "Signature error"}
+      _ -> {:error, :invalid_signature, "Signature error"}
+    end
+  end
+
+  defp check_expiration(payload, options) do
+    case payload do
+      %JOSE.JWT{fields: %{"exp" => exp, "iat" => iat}}
+      when is_integer(exp) and is_integer(iat) and exp > iat ->
+        if exp - iat > options[:max_lifetime] do
+          {:error, :too_long_lifetime, "Token lifetime is higher than expected"}
+        else
+          if exp > :os.system_time(:second) do
+            {:ok}
+          else
+            {:error, :token_expired, "Token expired"}
+          end
+        end
+
+      _ ->
+        {:error, :invalid_exp_iat_claim, "Bad exp or iat claim"}
+    end
+  end
+
+  defp check_audience(payload, options) do
+    case payload do
+      %JOSE.JWT{fields: %{"aud" => audience}} ->
+        if audience == options[:audience] do
+          {:ok}
+        else
+          {:error, :unknwon_audience, "Unkown audience"}
+        end
+
+      _ ->
+        {:error, :missing_aud_claim, "Missing aud claim"}
+    end
+  end
+
+  defp forbidden(conn, error_code, error_message) do
+    {:ok, response} =
+      Jason.encode(%{
+        errors: [
+          %{error_code: error_code, error_message: error_message}
+        ]
+      })
+
+    conn
```

Comment:
I trace very detailed error code for the sake of debuggability, but maybe I should add a condition to not return this very detailed error causes in production.